mempool/cat/reactor.go

```diff
@@ -0,0 +1,440 @@
+	}
+}
+
+// broadcastNewTx broadcast new transaction to all peers unless we are already sure they have seen the tx.
+func (memR *Reactor) broadcastNewTx(wtx *wrappedTx) {
+	msg := &protomem.Message{
+		Sum: &protomem.Message_Txs{
+			Txs: &protomem.Txs{
+				Txs: [][]byte{wtx.tx},
+			},
+		},
+	}
+	bz, err := msg.Marshal()
+	if err != nil {
+		panic(err)
+	}
+
+	for id, peer := range memR.ids.GetAll() {
+		if p, ok := peer.Get(types.PeerStateKey).(PeerState); ok {
+			// make sure peer isn't too far behind. This can happen
+			// if the peer is blocksyncing still and catching up
+			// in which case we just skip sending the transaction
+			if p.GetHeight() < wtx.height-peerHeightDiff {
+				memR.Logger.Debug("peer is too far behind us. Skipping broadcast of seen tx")
+				continue
+			}
+		}
+
+		if memR.mempool.seenByPeersSet.Has(wtx.key, id) {
+			continue
+		}
+
+		if peer.Send(mempool.MempoolChannel, bz) {
+			memR.mempool.PeerHasTx(id, wtx.key)
+		}
+	}
+}
+
+// requestTx requests a transaction from a peer and tracks it,
+// requesting it from another peer if the first peer does not respond.
+func (memR *Reactor) requestTx(txKey types.TxKey, peer p2p.Peer) {
+	if peer == nil {
+		// we have disconnected from the peer
+		return
+	}
+	memR.Logger.Debug("requesting tx", "txKey", txKey, "peerID", peer.ID())
+	msg := &protomem.Message{
+		Sum: &protomem.Message_WantTx{
+			WantTx: &protomem.WantTx{TxKey: txKey[:]},
+		},
+	}
+	bz, err := msg.Marshal()
+	if err != nil {
+		panic(err)
+	}
+
+	success := peer.Send(MempoolStateChannel, bz)
+	if success {
+		memR.mempool.metrics.RequestedTxs.Add(1)
+		requested := memR.requests.Add(txKey, memR.ids.GetIDForPeer(peer.ID()), memR.findNewPeerToRequestTx)
+		if !requested {
+			memR.Logger.Error("have already marked a tx as requested", "txKey", txKey, "peerID", peer.ID())
+		}
+	}
+}
+
+// findNewPeerToSendTx finds a new peer that has already seen the transaction to
+// request a transaction from.
+func (memR *Reactor) findNewPeerToRequestTx(txKey types.TxKey) {
+	// ensure that we are connected to peers
+	if memR.ids.Len() == 0 {
+		return
+	}
+
+	// pop the next peer in the list of remaining peers that have seen the tx
+	// and does not already have an outbound request for that tx
+	seenMap := memR.mempool.seenByPeersSet.Get(txKey)
+	var peerID uint16
+	for possiblePeer := range seenMap {
+		if !memR.requests.Has(possiblePeer, txKey) {
+			peerID = possiblePeer
+			break
+		}
+	}
+
+	if peerID == 0 {
+		// No other free peer has the transaction we are looking for.
+		// We give up 🤷‍♂️ and hope either a peer responds late or the tx
+		// is gossiped again
+		memR.Logger.Info("no other peer has the tx we are looking for", "txKey", txKey)
+		return
+	}
+	peer := memR.ids.GetPeer(peerID)
+	if peer == nil {
+		// we disconnected from that peer, retry again until we exhaust the list
+		memR.findNewPeerToRequestTx(txKey)
+	} else {
+		memR.mempool.metrics.RerequestedTxs.Add(1)
+		memR.requestTx(txKey, peer)
```

Comment:
What do you think about returning an error from the  `findNewPeerToRequestTx` to the caller to inform them of the state of the request (whether it's resolved or not)? right now, the request may be silently dropped and never get resolved (or it isn't the case?).
Also, in case of failure of `memR.requestTx(txKey, peer)`, it may be necessary to re-request the `txKey`.

Reply:
> Also, in case of failure of memR.requestTx(txKey, peer), it may be necessary to re-request the txKey.

This is currently how it works, if the requested tx isn't received in the timeout period, we classify it as failing and fire a callback which rerequests the same tx but to a different peer

Reply:
Thanks, it makes sense, can you please point me to where the timeout logic is implemented?